samples/nrf9160/serial_lte_modem/src/slm_at_icmp.c

```diff
@@ -0,0 +1,496 @@
+	data = ip_buf + header_len;
+	data[0] = ICMP_ECHO_REQ;                 /* Type (echo req) */
+	data[1] = 0x00;                          /* Code */
+	/* data[2..3] = checksum, calculated later */
+	data[4] = 0x00;                         /* Identifier */
+	data[5] = 0x00;                         /* Identifier */
+	data[6] = seqnr >> 8;                   /* seqnr */
+	data[7] = ++seqnr;                      /* seqr */
+
+	/* Payload */
+	for (int i = 8; i < total_length - header_len; i++) {
+		data[i] = (i + seqnr) % 10 + '0';
+	}
+
+	/* ICMP CRC */
+	calc_ics(data, total_length - header_len, 2);
+
+	rep = ICMP_ECHO_REP;
+
+	/* Send the ping */
+	do {
+		errno = 0;
+		delta_t = 0;
+		start_t = k_uptime_get();
+
+		fd = nrf_socket(NRF_AF_PACKET, NRF_SOCK_RAW, 0);
+		if (fd < 0) {
+			LOG_ERR("socket() failed: (%d)", -errno);
+			return (u32_t)delta_t;
+		}
+
+		ret = nrf_send(fd, ip_buf, total_length, 0);
+		if (ret <= 0) {
+			LOG_ERR("nrf_send() failed: (%d)", -errno);
+			goto close_end;
+		}
+
+		fds[0].handle = fd;
+		fds[0].requested = NRF_POLLIN;
+		ret = nrf_poll(fds, 1, ping_argv.waitms);
+		if (ret <= 0) {
+			LOG_ERR("nrf_poll() failed: (%d) (%d)", -errno, ret);
+			m_callback("#XPING: timeout\r\n");
+			goto close_end;
+		}
+
+reread:
+		len = nrf_recv(fd, ip_buf, NET_IPV4_MTU, 0);
+		if (len <= 0) {
+			LOG_ERR("nrf_recv() failed: (%d) (%d)", -errno, len);
+			goto close_end;
+		}
+		if (len < header_len) {
+			LOG_INF("nrf_recv() wrong data (%d)", len);
+			goto reread;
+		}
+
+		delta_t = k_uptime_delta(&start_t);
+
+		/* Check for ICMP EchoReply */
+		if (data[0] != rep) {
+			/* We got unsolicited message
+			 * drop it and wait for echo-reply
+			 */
+			goto reread;
+		} else {
+			if (ip_buf[IP_PROTOCOL_POS] != ICMP) {
+				/* Not ipv4 echo reply, ignore silently */
+				goto reread;
+			} else {
+				/* Check ICMP HCS */
+				hcs = check_ics(data, len - header_len);
+				if (hcs != 0) {
+					LOG_ERR("HCS error %d", hcs);
+					goto close_end;
+				}
+				/* Payload length */
+				pllen = (ip_buf[2] << 8) + ip_buf[3];
+			}
+
+			/* Check seqnr and length */
+			plseqnr = data[7];
+			if (plseqnr != seqnr) {
+				LOG_ERR("error sequence numbers %d %d",
+					plseqnr, seqnr);
+				goto close_end;
+			}
+			if (pllen != len) {
+				LOG_ERR("error length %d %d", pllen, len);
+				goto close_end;
+			}
+
+			/* Result */
+			sprintf(buf, "#XPING: %d.%03d\r\n",
+				(u32_t)(delta_t)/1000,
+				(u32_t)(delta_t)%1000);
+			m_callback(buf);
+		}
+
+close_end:
+		(void)nrf_close(fd);
+	} while (0);
+
+	return (u32_t)delta_t;
+}
+
+void ping_task(struct k_work *item)
+{
+	struct addrinfo *si = ping_argv.src;
+	struct addrinfo *di = ping_argv.dest;
+	u32_t sum = 0;
+	u32_t count = 0;
+
+	ARG_UNUSED(item);
+
+	for (int i = 0; i < ping_argv.count; i++) {
+		u32_t to = send_ping_wait_reply();
+
+		if (to > 0)  {
+			count++;
+			sum += to;
+		}
+		k_sleep(ping_argv.interval);
+	}
+
+	if (count > 1) {
+		u32_t ave = (sum + count/2) / count;
```

Comment:
avg is the common shorthand for average, not ave.

Reply:
updated